folium/features.py

```diff
@@ -416,6 +416,15 @@ def __init__(self, data, style_function=None, name=None,
             self.data = json.loads(json.dumps(data.__geo_interface__))  # noqa
         else:
             raise ValueError('Unhandled object {!r}.'.format(data))
+
+        callers = {'style_function': style_function, 'highlight_function': highlight_function}
+
+        for caller_key in callers:
+            caller = callers[caller_key]
+            if caller is not None:
```

Comment:
No need for the `None` and `caller = callers[caller_key]` and we can drop the assert here in lieu of Python Exceptions. You can use `.items()` like:

```python
for key, caller in callers.items():
   if not callable(caller) and not isinstance(caller(self.data['features'][0]), dict):
       raise ValueError(f'{key} must be a callable function returning a dictionary.')
```

Reply:
Nice, I always forget about `.items()` on dicts. Been spending more time in the JS world lately.

Don't we need the `not None` check though? The default values for `style_function` and `highlight_function` are `None`s right now - so they'll fail this check if unmodified.

Reply:
> Don't we need the `not None` check though?

`if not callable(caller)` should take care of that if I'm not mistaken. The `caller` will be `None` and will be `True` in `not callable(caller)`, but I believe the `and` should be an `or`. We want to raise if one or the other is `True`, right?

> Also, will our CI travis build environment accept the f-string usage? Last time I tried to commit one it caused a failure so I've been using `.format()` instead.

I wrote that quickly and I use f-strings everywhere, it was not meant to be a copy-n-paste review, sorry.

With that said, we will drop Python 2.7 support in the next release and we can use a f-string backport package for py35, so I'm OK if we start introducing them now. Let's just wait for @Conengmo opinion on this one :smile:

Reply:
> The caller will be None and will be True in not callable(caller), but I believe the and should be an or. We want to raise if one or the other is True, right?

But this will pass any other uncallable object, in addition to `None`, I'm pretty sure - like the dictionary in #1020.

IMO there should be only two clear passing cases for the callers - `None` and functions returning dicts.

Calling an uncallable (like `None(self.data['features'][0])`) raises `TypeError`. So we need the `and` operator to take the `False` from `callable` if our second call results in that `TypeError`, and still raise our custom `ValueError`.

So we need:

```
if not isinstance(caller, None):
    if not callable(caller) and isinstance(caller(self.data['features'][0]), dict):
        raise ValueError('Pass a function returning a dictionary to `{}`'.format(key))
```

Or maybe pack all the conditionals into one line? I know it is long, but I want to make it as airtight as possible and give useful feedback to the user. Let me know what you think.